designs/NuGet-ClientCertificates.md

```diff
@@ -0,0 +1,221 @@
+## Client certificates in NuGet
+
+Implements: https://github.com/NuGet/Home/issues/5773
+* Status: **Purposed**
+* Author(s): [Shkolka Volodymyr](https://github.com/BlackGad)
+
+## Issue
+
+[5773](https://github.com/NuGet/Home/issues/5773) - NuGet cannot restore from HTTPS sources that require Client Certificates
+
+## Background
+
+Developer tools that supports [Client Certificate Authentication](https://blogs.msdn.microsoft.com/kaushal/2015/05/27/client-certificate-authentication-part-1/).
+* [NPM](https://docs.npmjs.com/misc/config#cert)
+* [Maven](https://maven.apache.org/guides/mini/guide-repository-ssl.html)
+* [Teamcity](https://www.jetbrains.com/help/teamcity/using-https-to-access-teamcity-server.html#UsingHTTPStoaccessTeamCityserver-ConfiguringJVMforauthenticationwithclientcertificate)
+
+It is necessary to add client certificate authentication feature support to NuGet.
+
+## Who are the customers
+
+All .NET Core customers.
+
+## Requirements
+
+* Ability to specify client certificates from common [certificate stores](https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/working-with-certificates#certificate-stores)
+* Ability to specify client certificates from external files in standard [formats](https://en.wikipedia.org/wiki/X.509#Certificate_filename_extensions)
+* Ability to specify inline client Base64 encoded DER certificate in [PEM](https://en.wikipedia.org/wiki/Privacy-Enhanced_Mail) format
+* Ability to configure settings with command line
+
+## Solution
+
+1) Find specified in NuGet configuration file certificates (common implementation for [all products](https://github.com/NuGet/NuGet.Client/tree/dev/src/NuGet.Core/NuGet.Configuration) )
+2) Apply them prior any web request to appropriate http handler (product specific)
+
+## NuGet configuration changes
+
+Provide new configuration section `clientCertificates` which may have children of 3 types:
```

Comment:
I was busy last week and didn't have a chance to look at this again, but I quickly checked the existing client credentials configuration, and think client certs could fit in there: 

```xml
<configuration>
  <packageSourceCredentials>
    <source1>
      <add key='UserName' value='test_user' />
      <add key='ClearTextPassword' value='test_password' />
    </source1>
    <source2>
      <add key='CertStore' value='cert:\CurrentUser\My' />
      <add key='FindByThumbprint' value='4894671ae5aa84840cc1079e89e82d426bc24ec6' />
    </source2>
    <source3>
      <add key='CertFile' value='certificate.pfx' />
      <add key='ClearTextPassword' value='...' />
    </source3>
    <source4>
      <add key='CertPEM' value='-----BEGIN CERTIFICATE-----
MIIEjjCCA3agAwIBAgIJIBkBGf8AAAAPMA0GCSqGSIb3DQEBCwUAMIGzMQswCQYD
...
tJl1UvF7GWJd0yNyPVqCCnBY
-----END CERTIFICATE-----' />
    </source4>
  </packageSourceCredentials>
</configuration>
```

I don't think we have a schema for the nuget.config file, but if we did, this would avoid needing to change it. Certs that need passwords could use the existing password and clear text password to specify the cert password. There would be no ambiguity about when different sources need different certs (and it will be easier to avoid "leaking" cert info to sources that don't need client certs). 

I think this would be a better user experience. @nkolev92  @anangaur @rrelyea  what do you think?